### ISSUE TITLE
[css3 positioning] support position:sticky inside an overflow:hidden|auto on general parents

### ISSUE THREAD

**Issue by hunboy** [w3c/csswg-drafts#865]
css3 position

Currently the position:sticky element exclusively works when the all of general parents are overflow:visible. This is problematic a little whilst we use overflow:hidden trick for clearfix etc.

Testcase attached. Here is a simplesample, how is the layout broken when we change the overflow to visible, however sticky starts to work in that case. I detected, all of general parent-elements in the parent-path must be overflow: visible, which is weird a little. Any overflow:hidden in the parent-path kills the sticky.

ER: sync somehow the spec to allow sticky subcontainers inside a overflow:hidden general parent.

testcase: https://jsfiddle.net/utasir/rmmkxq62/11/

**Comment by hunboy** [comment]
mozbugzilla link : https://bugzilla.mozilla.org/show_bug.cgi?id=1329203

**Comment by pedi** [comment]
maybe not really a bug but by implementation. Check this specs out

http://www.coreyford.name/files/position-sticky-presentation/

> The box’s position depends on its containing block (established as for position:static) **as well as its scrolling container, defined by the nearest ancestor in the same document with a computed value for ‘overflow-x’ or ‘overflow-y’ other than ‘visible’,** or the viewport if no such ancestor exists.

**Comment by niutech** [comment]
The official spec for position: sticky is at [W3C](https://www.w3.org/TR/css-position-3/#sticky-pos) and there is no mention of `overflow: hidden`. I do not see a reason why an element with `position: sticky` cannot be sticky inside <body> with `overflow-x: hidden`.

**Comment by mstange** [comment]
This spec seems to have changed quite a bit since the implementation in Gecko. It no longer mentions scrolling ancestors, it exclusively talks about containing blocks now. I agree that the current writing of the spec would allow crossing overflow:hidden / auto descendants when computing the offset. But I don't know what the rationale behind that change to the spec was, or whether implementing that change is web compatible.

**Comment by sidhu663** [comment]
+1, I find this a very strange limitation especially when the element in question is not actually hidden. It makes it much more fragile to use it given that a change to a parent container / component could break sticky behaviour entirely

**Comment by valtlai** [comment]
It would be nice to make this work:

```html
<div class="table-wrapper">
	<table>
		<tr><th>Foo</th><th>Bar</th></tr>
		<tr><td>sin</td><td>cos</td></tr>
		<tr><td>tan</td><td>baz</td></tr>
		...
	</table>
</div>
```

```css
.table-wrapper {
	overflow-inline: auto;
}

th {
	position: sticky;
	inset-block-start: 0;
}
```

[**See demo**](https://codepen.io/valtlai/pen/XVrbPY)

I need both scrolling in the inline direction and a sticky header.

**Comment by SimplyPhy** [comment]
I'm building a dynamic and responsive table template in SCSS which requires this functionality in many situations.  Currently, everything works beautifully in Chrome [63] and Edge [41], but the sticky behavior fails [somewhat gracefully, thankfully] in Firefox [53] and Safari [11].

Here's a basic test case for overflow-x and overflow-y support:
https://codepen.io/SimplyPhy/pen/oEZKZo

**Comment by jonjohnjohnson** [comment]
@SimplyPhy was going to comment and say safari just needed prefix for sticky, but now I see you updated your comment. Though safari still has an issue 'sticking' your `td:first-child {top: 0}`, I'm guessing because it doesn't calculate the top edge of it's parent `tr` like other browsers. But if you remove that property declaration, I'm guessing the behavior is perfectly how you want it?

**Comment by SimplyPhy** [comment]
Yup, thanks @jonjohnjohnson and nice find.

**Comment by natematykiewicz** [comment]
I have the following rule:

```css
html, body {
  overflow-x: hidden;
}
```

and because of it, I'm not able to use `position: sticky;` at all. I'm trying to add some submit buttons on a form, and have them sticky to the bottom of the window `position: sticky; bottom: 0;` when outside the viewport. It works great and I love the feature, but this overflow-x rule prevents it from working. It sounds very similar to what @niutech mentioned.

**Comment by jonjohnjohnson** [comment]
@natematykiewicz though not ideal, I believe all current browser versions support sticky behavior if you alter your DOM structure so that the scrolling element is inside the body/html, filling out the screen. This does make it so browsers that resize the viewport on scrolling of the `document.scrollingElement` won't resize, such as ios safari. Notice the example code @SimplyPhy provided with for his table utility. Hope that helps until the spec is figured out.

**Comment by jonjohnjohnson** [comment]
Here's the most layman way to describe the implemenation if anybody is still confused...
> overflow: hidden[/auto] is not preventing position: sticky from working. But if you set overflow to hidden on any ancestor of your sticky element, then this ancestor element will be the scrolling container for your sticky element. - [https://stackoverflow.com/a/44929597/5891228](https://stackoverflow.com/a/44929597/5891228)

And the concept here relates to the issue @frivoal made about missing terminology for a possibly scrollable element that actually scrolls.

[#2322 [cssom-view][css-overflow] Missing terminology](https://github.com/w3c/csswg-drafts/issues/2322)

Not saying I think this is blocked by that, but wanted to make the connection nonetheless.

**Comment by natematykiewicz** [comment]
@jonjohnjohnson my previous comment was that I have `overflow-x: hidden;` on the body and html tag, and that prevents `position: sticky;` from working on anything. I did just realize that if I put the `overflow-x: hidden;` on only the body or the html (but not BOTH), then it works (this is in Chrome on a Mac). So, maybe I can actually adjust my CSS to get the best of both worlds. It's just odd that preventing the body and html from horizontally scrolling means that I can no longer get elements to stick to the top of the screen. I could see `overflow-y: hidden;` causing problems, but not `overflow-x: hidden;`.

**Comment by jonjohnjohnson** [comment]
@natematykiewicz 

The computed values of both `overflow-x` & `overflow-y` are...
> as specified, except with` visible` computing to `auto` if one of `overflow-x` or `overflow-y` is not `visible` - [css-overflow-3/#overflow-properties](https://www.w3.org/TR/css-overflow-3/#overflow-properties)

This is what causes our issue and is spec compliant.

Furthermore, check your "(but not BOTH)" solution in safari, including iOS, because that approach hasn't historically given people what they have sought out. -> [overflow-xhidden-doesnt-prevent-content-from-overflowing...](https://stackoverflow.com/questions/14270084/overflow-xhidden-doesnt-prevent-content-from-overflowing-in-mobile-browsers). Again, the most robust solution is to scroll an element WITHIN the body, not the body/html/document. Good luck.

**Comment by jonjohnjohnson** [comment]
Just wanted to make a note for others about how allowing general `overflow: hidden` ancestors to not be the "scrolling context" for a sticky element isn't always desirable. When using a modal view that programmatically turns off/on scrolling with `hidden`/`auto` a sticky element would jump around the screen between the swapping scrolling contexts.

Also, when we (eventually) have `overflow: clip` implemented everywhere, this should no longer be an issue?

- [#1971 - Is the box a scroll container if only one of 'overflow-x' or 'overflow-y' is 'clip'?](https://github.com/w3c/csswg-drafts/issues/1971)
- https://drafts.csswg.org/css-overflow-3/#valdef-overflow-clip

**Comment by sudheer496** [comment]
@SimplyPhy  in your codepen (https://codepen.io/SimplyPhy/pen/oEZKZo ) on table is there anyway to restrict scroll on header content. scroll looks odd on header.

**Comment by moritzjacobs** [comment]
For anybody stumbling upon this issue via google, here's a workaround (at least for my case) using a polyfill: https://stackoverflow.com/a/52420677/4903358

**Comment by protoEvangelion** [comment]
For my use case, one of the ancestors in the DOM had `overflow-x: hidden;`. _My goal was to simply hide horizontal scrollbar, so this may not apply to your use case_.

The way I fixed it was removing `overflow-x: hidden;` from that element in the DOM and adding it to body instead!

```css
body {
  overflow-x: hidden;
}
```

**Comment by damienroche** [comment]
@protoEvangelion I realized after trying that the solution does not work on mobile browsers :(

**Comment by jonjohnjohnson** [comment]
@damienroche I believe androids webview behaves according to spec here, but it's iOS doesn't. If you want to chip in and voice concern about their noncompliance here's the public ticket.

[WebKit Bugzilla - Bug 153852 - <body> with overflow:hidden CSS is scrollable on iOS](https://bugs.webkit.org/show_bug.cgi?id=153852)

**Comment by OliverJAsh** [comment]
> Also, when we (eventually) have `overflow: clip` implemented everywhere, this should no longer be an issue?

It does sound like `overflow: clip` is the answer here. What is the current position of this standard? I don't see any browsers implementing it. 🤔

**Comment by emilio** [comment]
Gecko implements `overflow: -moz-hidden-unscrollable` which is effectively `overflow: clip`. Don't know the background on why it's not unprefixed.

**Comment by Baedda** [comment]
The fact that `overflow: hidden;` on an ancestor makes this container become the new scrolling container for the sticky element, takes away the opportunity to clip the sticky container in any way.

I have a use case where the height of the sticky element and the height of its parent are unknown. I want the sticky element to be sticky on browser scroll. But I dont want the sticky element to overflow its parent. With the current spec this seems to be impossible. But I believe that this is a common use case.

**Comment by joma74** [comment]
No clue for in combination with scroll-snap on parent. Page with 100vh per section, sticky as dynamic css for header nav part of first section overflows the scroll bar.

**Comment by WangXBruc** [comment]
> It would be nice to make this work:
> 
> ```
> <div class="table-wrapper">
>     <table>
>         <tr><th>Foo</th><th>Bar</th></tr>
>         <tr><td>sin</td><td>cos</td></tr>
>         <tr><td>tan</td><td>baz</td></tr>
>         ...
>     </table>
> </div>
> ```
> 
> ```css
> .table-wrapper {
>     overflow-x: auto;
> }
> 
> th {
>     position: sticky;
>     top: 0;
> }
> ```
> 
> [**See demo**](https://codepen.io/valtlai/pen/XVrbPY/left/)
> 
> I need both horizontal scrolling and sticky header.

Hello. Have you found the solution? I have the same require with yours

**Comment by valtlai** [comment]
@WangXBruc Unfortunately not.

**Comment by jonjohnjohnson** [comment]
@valtlai @WangXBruc I don't exactly condone this fix, but if you really need some sort of nested separate scroller with separate axis treatment, this is the best approach yet.
https://uxdesign.cc/position-stuck-96c9f55d9526

@OliverJAsh @emilio Gecko bug for renaming/implementing `overflow: clip`.
https://bugzilla.mozilla.org/show_bug.cgi?id=1531609

@Baedda Tangentially...
> The fact that overflow: hidden; on an ancestor makes this container become the new scrolling container for the sticky element, takes away the opportunity to clip the sticky container in any way.

There are other ways to "clip" the sticky container, by using `clip-path`, even though the "overflow" can affect the geometry of scrollable content, it still behaves well in most cases. 

@joma74 
> No clue for in combination with scroll-snap on parent. Page with 100vh per section, sticky as dynamic css for header nav part of first section overflows the scroll bar.

You can use `calc()` to offset a "fullscreen" height with a sticky nav. But some browsers (iOS safari) don't report `100vh` as the height of the unscrolled viewport. (https://nicolas-hoizey.com/2015/02/viewport-height-is-taller-than-the-visible-part-of-the-document-in-some-mobile-browsers.html)

@Baedda @joma74 check these demos for examples of sticky/clip-path/calc()/vh
- https://jsbin.com/tedowel/edit
- https://jsbin.com/lefekiv/edit
- https://jsbin.com/subojom/edit

**Comment by joshjg** [comment]
@jonjohnjohnson The link you posted (https://uxdesign.cc/position-stuck-96c9f55d9526) requires javascript (basically, attaching a scroll listener which updates the positioning of other elements). I don't think this is an acceptable workaround, especially from a performance perspective.

**Comment by JacobDB** [comment]
@joshjg I agree &ndash; it's a solution, but given that `position: sticky` is a native CSS feature, we shouldn't need to rely on JavaScript for this use case.

**Comment by jonjohnjohnson** [comment]
@joshjg @JacobDB I provided that link in response to a [case](https://github.com/w3c/csswg-drafts/issues/865#issuecomment-350585274) which is a bit different/more than this filing, so I'd suggest opening up a separate issue. Something along the lines of...

 **"[css-positioned-3] expand sticky logic to allow sticky positioning behavior between nested scrollers, specifically cross axis"**

Especially since it seems a lot of folks have shown support for the use  [case](https://github.com/w3c/csswg-drafts/issues/865#issuecomment-350585274) @valtlai provided.

**Comment by BenMorel** [comment]
I want to emphasize my support for @valtlai's request as well:

```css
.table-wrapper {
    overflow-x: auto;
}

th {
    position: sticky;
    top: 0;
}
```

Currently you have to choose between having the table scroll horizontally, or sticking the headers. You can't have both, which is a shame.

The only "solution" that does *not* involve JavaScript is forcing the table to a given height, which is a really ugly and counter-user-friendly solution with its 2 scrollbars, one for the page and one for the table (especially when the table is the central UI element in the page).

**Comment by natematykiewicz** [comment]
I tried to do sticky headers on a table once for a nice UX improvement. I gave up for the very reason you just said. The table was in a div to allow for for horizontal scrolling and therefore could not have its headers stuck...

**Comment by Saw-mon-and-Natalie** [comment]
Looks like if you set `overflow` to a non-default value for the sticky element ancestor, you would need to set the ancestor height as well. Otherwise, the `position: sticky` would not work. For example:

```css
.wrapper {
    overflow-x: hidden;
    height: 100vh;
}
```

This works on the `chrome 78 (64-bit)` on `Win 7`.
Example: [codepen](https://codepen.io/sawmon_and_natalie/full/qBBJvXY)

**Comment by er-parag91** [comment]
> > It would be nice to make this work:
> > ```
> > <div class="table-wrapper">
> >     <table>
> >         <tr><th>Foo</th><th>Bar</th></tr>
> >         <tr><td>sin</td><td>cos</td></tr>
> >         <tr><td>tan</td><td>baz</td></tr>
> >         ...
> >     </table>
> > </div>
> > ```
> > 
> > 
> > ```css
> > .table-wrapper {
> >     overflow-x: auto;
> > }
> > 
> > th {
> >     position: sticky;
> >     top: 0;
> > }
> > ```
> > 
> > 
> > [**See demo**](https://codepen.io/valtlai/pen/XVrbPY/left/)
> > I need both horizontal scrolling and sticky header.
> 
> Hello. Have you found the solution? I have the same require with yours

I was facing same issue, I did try number of CSS solution and finally it works the way you guys are looking for. Here is working codesandbox URL: 

https://codesandbox.io/embed/objective-frog-m4imr?fontsize=14&hidenavigation=1&theme=dark

And I am using react table however CSS styling is pretty easy to understand and Don't thing you would face any issue copying over solution even though you don't know react.

**Comment by BenMorel** [comment]
@er-parag91 It looks like your code is an example of the "solution" I mentioned above:

> The only "solution" that does not involve JavaScript is **forcing the table to a given height**, which is a really ugly and counter-user-friendly solution with its 2 scrollbars, one for the page and one for the table (especially when the table is the central UI element in the page).

You're forcing the table to `50vh` height. If you have a scrollable page, you end up with 2 scrollbars: one for the page, and one for the table. Ugly if you ask me!

**Comment by melihoeztuerk** [comment]
Facing same issue. Well seems like this will still not be supported:
> Any ancestor between the sticky element and its user-scrollable container with overflow computed as anything but visible/clip will effectively prevent sticking (https://caniuse.com/#feat=css-sticky)

I would prefer to search for a nice workaround instead of a "clean" solution.

**Comment by tabatkins** [comment]
As far as I can tell, this request is based on using `overflow:hidden` to force a containing to become a BFC so it'll clearfix.

Since we (a) have better ways to force a BFC now that don't interact badly with `position:sticky` (such as `display: flow-root;`), and (b) rarely if ever need to clearfix now (since that's mostly an artifact of float-based layouts), I think we can safely close this as no change.

**Comment by coreyworrell** [comment]
@tabatkins hmm, you may have missed the multiple comments about sticky headers on horizontally scrollable tables (a common method for handling large tables on smaller devices). There is also the issue of the common case of having a site wrapper that hides any `overflow-x` for off screen navigation and similar animation techniques. That makes it impossible to use sticky elements.

**Comment by shrpne** [comment]
@tabatkins in my practice `overflow` is rarely used for BFC, BFC is a thing from the past, when layout was made with `float`, for now, we have `flex` or even `grid` and developers have no reason to use `float` and care about BFC.

I agree with @coreyworrell, 90% of my websites have `overflow` wrapper to hide offscreen things like protruding images.

**Comment by JacobDB** [comment]
I agree &endash; I typically set `overflow: hidden;` on a container to prevent my column system from overflowing. A very simplified example is below.

```scss
.row {
    display: flex;
    width: calc(100% + $column_gap);
}

.col {
    flex: 1 1 25%;
    padding-right: $column_gap;
}
```

This method keeps everything aligned just right with an accurate gap between each element, but of course means that the `.row` almost always overflows the width of its containers. To combat this, I set `.page-wrapper { overflow: hidden; }`, which works great, except when I need to use something with `position: sticky;`.

Admittedly, my use case could be solved by switching to `grid` instead of `flex`, but in my [testing](https://codepen.io/JacobDB/pen/rNBdxOg), I've been unable to replicate the features I need using `grid`.

**Comment by tabatkins** [comment]
> hmm, you may have missed the multiple comments about sticky headers on horizontally scrollable tables

Hmm, I had. I see the issue now.

It seems like the case here is wanting to put horizontal scrollbars on a too-wide table, but *not* put vertical scrollbars on it when it's taller than 100vh. (Or generally "whatever the height of the screen is", since that's a... difficult-to-predict value on mobile.) And then still get the header rows to be sticky as the table scrolls off, but stay contained to the table itself as normal.

Okay, I'll re-open for this issue.

**Comment by tabatkins** [comment]
For the "I'm just using an `overflow:hidden` wrapper to prevents things from visually spilling out" case, `overflow:clip` achieves that without making it scrollable, and so *shouldn't* affect stickyness. That value isn't implemented in Chrome yet, but `-moz-hidden-unscrollable` is a prefixed name for the value, and if I'm reading previous comments correctly, it does indeed work with sticky (doesn't block stickiness). So I'll consider that part of the issue solved.

(The naming of "hidden" is unfortunate; really it's an "I'm handling scrolling with JS, don't worry about it" value, but it claimed a useful name that makes it sound more general than it is. What we're now calling "clip" should have been named "hidden". ^_^)

**Comment by saschanaz** [comment]
>The naming of "hidden" is unfortunate; really it's an "I'm handling scrolling with JS, don't worry about it" value, but it claimed a useful name that makes it sound more general than it is.

Maybe a new alias `overflow: manual`? 🤔

**Comment by tabatkins** [comment]
> but not put vertical scrollbars on it when it's taller than 100vh. (Or generally "whatever the height of the screen is", since that's a... difficult-to-predict value on mobile.)

Expanding on this, is the problem just that a `max-height` on a table is hard to predict on mobile? (Because `100vh` might be too large depending on what UI is being shown.) If that's the case, would #4329 solve the issue?

**Comment by coreyworrell** [comment]
> Expanding on this, is the problem just that a `max-height` on a table is hard to predict on mobile? (Because `100vh` might be too large depending on what UI is being shown.) If that's the case, would #4329 solve the issue?

@tabatkins I think you said it correct before:

> It seems like the case here is wanting to put horizontal scrollbars on a too-wide table, but not put vertical scrollbars on it when it's taller than 100vh. (Or generally "whatever the height of the screen is", since that's a... difficult-to-predict value on mobile.) And then still get the header rows to be sticky as the table scrolls off, but stay contained to the table itself as normal.

Adding a `max-height` can cause double scroll bars (one for window, one for table). It is preferable to allow the table it's auto height.

**Comment by jonjohnjohnson** [comment]
@tabatkins & @coreyworrell I think you are talking about what I addressed in https://github.com/w3c/csswg-drafts/issues/865#issuecomment-488522567.

> "[css-positioned-3] expand sticky logic to allow sticky positioning behavior between nested scrollers, specifically cross axis"

(And honestly, there is a strange president for nested sticky logic that shares a containing block, see http://output.jsbin.com/zeyanef/7, best in webkit/firefox. Though, this example doesn't share the containing block through a scroll container, but instead through sticky ancestors.)

This seems to be the only remaining issue in the thread and should simply be broken out, allowing this thread to be closed. Because we have `overflow:clip` solving the general ancestor between the sticky element and it's scroll container (containing block) issue. And browsers have now consolidated on a solution for `scrollingElement` weirdness with horizontal overflow between the `html` & `body` elements, where an author just needs to make sure `body` isn't an intermediate scroll container between the sticky element and the `html` `scrollingElement`.

Final note, it really would be a bad idea to retroactively make `overflow:hidden` elements do what most in this thread desire for the reasons I stated in https://github.com/w3c/csswg-drafts/issues/865#issuecomment-387908285.

> When using a modal view that programmatically turns off/on scrolling with hidden/auto a sticky element would jump around the screen between the swapping scrolling contexts.

This shows exactly why the specified behavior of `overflow:clip` is what is really needed, it just needs to get implemented everywhere. Elements with `overflow:hidden` are still scrolling elements even if only programmatically.

**Comment by shuklaayush** [comment]
Is there a workaround for what @valtlai mentioned (https://github.com/w3c/csswg-drafts/issues/865#issuecomment-350585274)? The link that @jonjohnjohnson shared does not seem to work for native html tables. I too don't want to add a fixed height since the extra scrollbar makes things ugly.

**Comment by johannesodland** [comment]
> For the "I'm just using an `overflow:hidden` wrapper to prevents things from visually spilling out" case, `overflow:clip` achieves that without making it scrollable, and so _shouldn't_ affect stickyness. That value isn't implemented in Chrome yet, but `-moz-hidden-unscrollable` is a prefixed name for the value, and if I'm reading previous comments correctly, it does indeed work with sticky (doesn't block stickiness). So I'll consider that part of the issue solved.

@tabatkins: Chrome has closed their issue as won't fix: https://bugs.chromium.org/p/chromium/issues/detail?id=566226

Maybe we should push for them to reopen the issue?

**Comment by Fuseteam** [comment]
for anyone looking for a solution for native html table i have posted on stackoverflow here:https://stackoverflow.com/questions/63384685/overflowauto-doesnt-work-as-expected-with-positionsticky and also a jsfiddle here: https://jsfiddle.net/uf468ocy/
basically it  has a copy table with only the headers which is shown as the header is scrolled out of position.

it possible way i was thinking to make it work better to have ti work natively i would think to put sticky op the scrolling container with overflow such that when the container is scrolled out of view the top "sticks" while the actual container continues to scroll.......if that makes sense

for example say i have a div #data like this

```
<div id="data" style="overflow:auto;position:sticky;top:0;">
<div id="sticky-element" style="position:sticky;top:0;">
</div>
</div>
```
then we scrolling would stick data to the top of viewport instead of scrolling it past, while the content of data is then scolled which sticks the element at the top of the data which is the top of the screen or something like that

curremtly if the sticky element is larger than the view port it just scrolls past

**Comment by payalord** [comment]
How about this solution: https://uxdesign.cc/position-stuck-96c9f55d9526

**Comment by konkit** [comment]
> How about this solution: https://uxdesign.cc/position-stuck-96c9f55d9526

AFAIK, it assumes, that columns have a constant width. Otherwise, if a wider content change the proportions of columns in the table, then headers can misalign with the content.

**Comment by notrealdev** [comment]
Why this simple rule is bug???

**Comment by Fuseteam** [comment]
@notrealdev please read the context of this issue

**Comment by chrishtr** [comment]
> Maybe we should push for them to reopen the issue?

Good news: `overflow:clip` work is in progress in Chromium.

**Comment by WickyNilliams** [comment]
I hate to add to a long thread just to echo earlier thoughts, but I am also wanting a table with sticky headers and horizontal scrolling (and no explicit height).

My specific use-case is a comparison table. Sticky headers are help orient the user as they scroll through a potentially large table, ensuring they do not get lost. Horizontal scrolling is required because you want to maintain standard columnar layout (i.e. not use other responsive "tricks" like collapsing rows/columns), allowing for quick comparison between rows. Explicit height is not a suitable workaround as it relies on fragile/performance-intensive scripting. 

Anything less than that is a sub-optimal experience for users. @valtlai's codepen perfectly demonstrates the current issue https://github.com/w3c/csswg-drafts/issues/865#issuecomment-350585274

**Comment by hunboy** [comment]
WickyNilliams: i think you want this:

[https://jsfiddle.net/utasir/odrg9etx/](url)

**Comment by WickyNilliams** [comment]
@hunboy Thanks. But you may be surprised to learn that changing the display value of a table actually strips it of its semantics. See here: https://developer.paciellogroup.com/blog/2018/03/short-note-on-what-css-display-properties-do-to-table-semantics/. 

You can verify this in the dev tools. With `display: grid`, the accessibility tree for the table consists of `gridcell`s instead of `row`s.

So this is not suitable approach. (aside: in my use-case I do not care so much about the first column being sticky, just the headers, combined with overflow-x)

**Comment by hunboy** [comment]
Well, for your further questions you can find me on IRC at  ircs://freenode/%23%23learncss.

I opened this issue 4years ago based on those common usage we had at that time on the certain webpages.
After 4years we have now `display: flow-root` in the spec, which solves this clearfix problem in almost all of cases.
And my original tescase also works with flow-root: [https://jsfiddle.net/utasir/rmmkxq62/170/](url)
So this issue is about to close

**Comment by WickyNilliams** [comment]
OK, glad to hear you got it solved.

@tabatkins you acknowledged the outstanding issue with tables/scrolling here: https://github.com/w3c/csswg-drafts/issues/865#issuecomment-621273538. In light of that, is there a more specific, open issue about this particular use-case where discussion can continue? My search didn't turn up much

**Comment by ZhaoTim** [comment]
> https://codepen.io/SimplyPhy/pen/oEZKZo

I find when I scroll vertically, will trigger horizontally scroll......inside safari

**Comment by nowyuser1** [comment]
> I have the following rule:
> 
> ```css
> html, body {
>   overflow-x: hidden;
> }
> ```
> 
> and because of it, I'm not able to use `position: sticky;` at all. I'm trying to add some submit buttons on a form, and have them sticky to the bottom of the window `position: sticky; bottom: 0;` when outside the viewport. It works great and I love the feature, but this overflow-x rule prevents it from working. It sounds very similar to what @niutech mentioned.

I've tried for html overflow-x: hidden; and for body overflow-x: clip; 
It works fine for me also overflow clip is now supported for most of browsers

**Comment by Fuseteam** [comment]
@alexander-schranz that breaks in the case of overflow:auto, it is no longer able to scroll when needed https://jsfiddle.net/p3wxzngL/2/

**Comment by kientran-clubspark** [comment]
test

**Comment by chenggwang** [comment]
[e.g:  Locate the head element of the alphabet](https://developer.mozilla.org/zh-CN/docs/Web/CSS/position).Why does it look like the letter A is pushed away by the letter C，Instead of covering？The sticky element will be "fixed" to the nearest ancestor with a "rolling mechanism"，I think it should be fixed when the threshold is reached. Later coverage comes first。Is it fixed to the nearest parent, whether scrolling or not？

**Comment by meloseven** [comment]
这是来自QQ邮箱的假期自动回复邮件。您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。

**Comment by chenggwang** [comment]
> [e.g: Stick the alphabet head element](https://developer.mozilla.org/zh-CN/docs/Web/CSS/position).Why does it look like the letter A is pushed away by the letter C，Instead of covering？The sticky element will be "fixed" to the nearest ancestor with a "rolling mechanism"，I think it should be fixed when the threshold is reached. Later coverage comes first。Is it fixed to the nearest parent, whether scrolling or not？

**Comment by shuishuidelin** [comment]
> 让它工作会很好：
> 
> ```
> <div class="table-wrapper">
>     <table>
>         <tr><th>Foo</th><th>Bar</th></tr>
>         <tr><td>sin</td><td>cos</td></tr>
>         <tr><td>tan</td><td>baz</td></tr>
>         ...
>     </table>
> </div>
> ```
> 
> ```css
> .table-wrapper {
>     overflow-x: auto;
> }
> 
> th {
>     position: sticky;
>     top: 0;
> }
> ```
> 
> [**看演示**](https://codepen.io/valtlai/pen/XVrbPY/left/)
> 
> 我需要水平滚动和粘性标题。

**Comment by kuohuanhuan** [comment]
`clear: both;` works for me.

**Comment by ObjectAssign** [comment]
Oh gawd it makes me so sad to see this is still not possible.. the crazy websites that could be made with this, honestly..

**Comment by meloseven** [comment]
这是来自QQ邮箱的假期自动回复邮件。您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。

**Comment by ObjectAssign** [comment]
> clear: both;

Works for you how? come with an example please :)

**Comment by nareshkvs** [comment]
> It would be nice to make this work:
> 
> ```
> <div class="table-wrapper">
>     <table>
>         <tr><th>Foo</th><th>Bar</th></tr>
>         <tr><td>sin</td><td>cos</td></tr>
>         <tr><td>tan</td><td>baz</td></tr>
>         ...
>     </table>
> </div>
> ```
> 
> ```css
> .table-wrapper {
>     overflow-x: auto;
> }
> 
> th {
>     position: sticky;
>     top: 0;
> }
> ```

Did you find any solution for this?

**Comment by AMUCOMO** [comment]
Well as told here (https://www.gatlin.io/blog/post/stuck-on-position-sticky) the very sad cross axis problem is indeed very unfortunate.  Especially considering this (https://stackoverflow.com/questions/6421966/css-overflow-x-visible-and-overflow-y-hidden-causing-scrollbar-issue).

**Comment by emilio** [comment]
> Well as told here (https://www.gatlin.io/blog/post/stuck-on-position-sticky) the very sad cross axis problem is indeed very unfortunate. Especially considering this (https://stackoverflow.com/questions/6421966/css-overflow-x-visible-and-overflow-y-hidden-causing-scrollbar-issue).

That seems like it'd be fixed by using overflow-y: clip instead, which doesn't cause the creation of a scroll containers.

**Comment by meloseven** [comment]
这是来自QQ邮箱的假期自动回复邮件。您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。

**Comment by AMUCOMO** [comment]
> > Well as told here (https://www.gatlin.io/blog/post/stuck-on-position-sticky) the very sad cross axis problem is indeed very unfortunate. Especially considering this (https://stackoverflow.com/questions/6421966/css-overflow-x-visible-and-overflow-y-hidden-causing-scrollbar-issue).
> 
> That seems like it'd be fixed by using overflow-y: clip instead, which doesn't cause the creation of a scroll containers.

Sadly not. At least for Firefox and Edge `  overflow: auto clip;
}` gets converted to `  overflow: auto hidden;
}`

**Comment by emilio** [comment]
Well, that is working as defined in https://drafts.csswg.org/css-overflow/#overflow-properties:

> The [visible](https://drafts.csswg.org/css-overflow/#valdef-overflow-visible)/[clip](https://drafts.csswg.org/css-overflow/#valdef-overflow-clip) values of [overflow](https://drafts.csswg.org/css-overflow/#propdef-overflow) compute to [auto](https://drafts.csswg.org/css-overflow/#valdef-overflow-auto)/[hidden](https://drafts.csswg.org/css-overflow/#valdef-overflow-hidden) (respectively) if one of [overflow-x](https://drafts.csswg.org/css-overflow/#propdef-overflow-x) or [overflow-y](https://drafts.csswg.org/css-overflow/#propdef-overflow-y) is neither visible nor clip.

but the link you pasted was about `x: visible` and `y: hidden` unnecessarily creating scrollbars, which can be prevented by using `x: visible` and `y: clip`.

**Comment by AaronBeaudoin** [comment]
This feature is useful for more than just horizontal scrolling on tables.

In my case the problem is combining rounded corners with sticky group headers inside a panel. Right now, the easiest way to give a panel/element rounded corners is to do `border-radius: <whatever>px` combined with `overflow: hidden` to prevent child elements with backgrounds from bleeding over the rounded areas at the corners. But that then prevents me from using `position: sticky` on any elements inside the panel.

My solution has been to not use `border-radius` or `overflow-hidden` and instead use a `clip-path: inset(0 0 0 0 round <whatever>px)` on the panel, but that comes with a new issue in my case (a bug with Chrome that seems to only occur when I'm using `clip-path` on huge areas such as panels that comprise most of the page) so `clip-path` is not an option either. Right now it's looking like I'm going to have to just use JS and avoid CSS `position: sticky` completely, which is really sad.

**Comment by DavidJCobb** [comment]
`clip-path` is unsuitable for other reasons as well. I recently got subpixel rounding errors that caused a browser (don't recall whether it was Firefox or Chrome) to clip a 1px border off of an element, even though that border should have fallen within the specified path.

It's incredibly unfortunate that setting `overflow-x: auto` on an ancestor causes `position: sticky; top: 0` to become locked to that ancestor's always-zero scroll Y-position.

**Comment by fantasai** [comment]
OK, @tabatkins and I filed a new issue about having an element be sticky to multiple scrollports: https://github.com/w3c/csswg-drafts/issues/8286

As far as we can tell, the rest of this issue should be handled by appropriate use of either `display: flow-root` or `overflow: clip`. @hunboy, @jonjohnjohnson, can you confirm?

**Comment by jonjohnjohnson** [comment]
@fantasai confirmed! The discussion in this issue has become so broad that it's hard for newcomers to easily understand which aspects are and are not outstanding.

**Comment by evgeniy-kolmak** [comment]
The only thing that helped for my detection `overflow: clip`. I was able to make friends with `postion: sticky`

**Comment by meloseven** [comment]
这是来自QQ邮箱的假期自动回复邮件。您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。

**Comment by akatuku** [comment]
I concur `overflow: clip` is helpful. For my use case, I created a custom scroll bar that mimics native scroll behavior it is doable(use only vanilla JS). I did notice some perf issues, but throttling and debouncing did help with perf.

**Comment by meloseven** [comment]
这是来自QQ邮箱的假期自动回复邮件。您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。

**Comment by towerofnix** [comment]
Hey, just commenting since we have a similar situation as @AaronBeaudoin brought up in https://github.com/w3c/csswg-drafts/issues/865#issuecomment-1267567002 but with some extra nuance that makes `overflow: hidden` appealing, and not a detail I've seen brought up here yet.

One of the greater benefits of `overflow: hidden` is that it clips "normally" positioned children, *but not* children with `position: absolute`. In other words you can use it to cut off corners of static children - or, in principle, sticky children! - that "seep out" past the rounded corners of the parent. But you're left with flexibility: if you do need other children to leave that containing box, *while still being placed relatively,* `position: absolute` is your go-to. (In a `position: relative` parent of course.)

We're taking advantage of basically this `overflow: hidden` behavior for hover-tooltips beneath links (in addition to cleaning the rounded corners of the containing panel):

<img width="283" alt="Floaty-looking box positioned neatly beneath a hovered link in ordinary text content, on top of other text content." src="https://github.com/w3c/csswg-drafts/assets/9948030/d1d78dd9-5925-4a51-80ac-9add8eba4959">

But in practice we aren't able to use `overflow: hidden`, of course, because it prevents the sticky heading element from scrolling with the page. So instead we're using `contain: paint` at the moment (`overflow: clip` is fine too, but worse browser support, and the same issue ahead). And this works fine... *until* we get to that case where `overflow: hidden`'s true power shines... and `contain: paint` sure doesn't!

<img width="240" alt="A rather tall tooltip near the bottom of a page. Where the main panel cuts off and the footer begins, the tooltip gets chopped in half!" src="https://github.com/w3c/csswg-drafts/assets/9948030/aa832c8c-a1dd-415c-a463-f1ca3657c525">

It seems like the only workable solution here is to set `overflow-y: scroll` on the main element, but this seriously alters and conflicts with the layout of the site (you're supposed to scroll the *page,* not the main content panel).

We could position the tooltips with JavaScript and separate them outright from the main page layout tree, but it's such a janky solution. Placement in the HTML tree, statically generated, and positioned relatively to a parent `position: relative` link isn't something we're interested in leaving behind, but it's not clear what current solutions *don't* involve using `contain: paint` or similar - and that's not compatible with `position: absolute` elements that sometimes extend past the containing element's border.

**Comment by JacobDB** [comment]
@towerofnix unfortunate that that's not the case, I actually ran in to similar situation recently, though not related to `position: sticky;`; I have a slideshow that needs `overflow: hidden;` to crop off its sibling slides, but navigation buttons that need to escape the container. There's workarounds of course, with additional containers, but it'd be awesome to have a way to "escape" the overflow property somehow when needed.

**Comment by webark** [comment]
I would say, that for 95% of the usecases that run into this, they would be fine falling back onto using the window for placement. Perhaps just a "sticky-window" or something like that additional, could solve the issue, but not open the need to select the anhoring ansestor. You would just get your parent, or the window. And if you choose the window, things like overflow and the ancestors scroll position is not longer taken into account, cause it can just be relative to the window.

**Comment by AxSamyn** [comment]
While I did not found a working solution to keep both overflow and sticky feature, I explored the possibility to disable overflow only when working with sticky elements.

The following piece of code resets the overflow property of elements that has a child marked with the sticky class.

```
*:has(:is(.sticky-top, .sticky-bottom)) {
    overflow: visible !important;
  }
```

I'm used to working with the bootstrap framework, so I based my css selector on the .sticky-top and .sticky-bottom classes, but feel free to change it to your needs.

Also, [I read](https://stackoverflow.com/a/63780286/4456593) that the sticky feature would only cause troubles with parents that have overflow with no fixed width or height set. So I explored further the possibility to disable only these elements.
The closest I could get was using the following piece of code :

```
  :is([style*="width: auto"], [style*="height: auto"]):has(:is(.sticky-top, .sticky-bottom)) {
    overflow: visible !important;
  }
  ```
  
  This requires the width and height attributes to be set on the elements and not via classes, which is quiet common... so I admit that this attempt may be worthless but might give you some ideas.
  
Let me know if I'm heading wrong way ;)

**Comment by flackr** [comment]
I propose that we change `position: sticky` to track both of the nearest scrolling containers per axis that "scroll" the content in that axis, i.e. not whether they have overflow in the axis, but whether the `overflow` value for that axis is `hidden | auto | scroll`, which in spec language may mean saying something like a box can be a horizontal scrolling container but not a vertical scrolling container.

This would solve the issue for tables which need to scroll in the other direction - see https://github.com/w3c/csswg-drafts/issues/865#issuecomment-350585274. Given the resolution we have taken for #12289 we can make this work by skipping over clip containers for that axis. E.g. if you set `.table-wrapper { overflow-x: auto; overflow-y: clip; }` then the element which is sticky position to the top (or bottom) could find the nearest ancestor **vertical** scrolling container making this use case work nicely. Here's a demo of the effect on a table header using a scroll driven animation: https://codepen.io/flackr/pen/WbwzEBo

![Image](https://github.com/user-attachments/assets/43d7a81d-0c93-409a-89d3-abe6b3484dfe)

This doesn't require any major spec changes, and should be relatively compatible as allowing `overflow: hidden | scroll | auto` and `overflow: clip` in the other axis is a new thing that has not yet been supported.

The issue in the OP (test case https://jsfiddle.net/utasir/rmmkxq62/11/ ) was taking advantage of the fact that overflow: hidden also establishes a new block formatting context, where overflow: clip [does not](https://www.w3.org/TR/css-overflow-3/#valdef-overflow-clip). There are other ways to accomplish the same and solve the original issue using overflow: clip which is a much better practice for content which is not supposed to be scrollable. For example, I forked the demo and was able to get the correct layout using `contain: layout` to ensure that all of the content in the article is in a new context: https://jsfiddle.net/zwd3fer2/4/

**Comment by meloseven** [comment]
这是来自QQ邮箱的假期自动回复邮件。您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。

**Comment by meloseven** [comment]
这是来自QQ邮箱的假期自动回复邮件。您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。
